jicoco/src/main/java/org/jitsi/xmpp/mucclient/MucClient.java

```diff
@@ -549,8 +561,67 @@ public String getId()
      */
     void stop()
     {
-        mucs.values().forEach(MucWrapper::leave);
-        xmppConnection.disconnect();
+        if (this.connectRetry != null)
+        {
+            this.connectRetry.cancel();
+        }
+
+        // If we are still not connected leave and disconnect my through
+        // errors
+        try
+        {
+            mucs.values().forEach(MucWrapper::leave);
+        }
+        catch(Exception e)
+        {
+            logger.error("Error leaving mucs", e);
+        }
+
+        try
+        {
+            xmppConnection.disconnect();
+        }
+        catch(Exception e)
+        {
+            logger.error("Error disconnecting xmpp connection", e);
+        }
+    }
+
+    /**
+     * The callable returned by this method describes the task of connecting
+     * and login to XMPP service.
+     *
+     * @return <tt>Callable<Boolean></tt> which returns <tt>true</tt> as long
+     *         as we're failing to connect.
+     */
+    private Callable<Boolean> getConnectAndLoginCallable()
+    {
+        return () ->
+        {
+            try
+            {
+                xmppConnection.connect();
+            }
+            catch(Exception t)
+            {
+                logger.warn(MucClient.this + " error connecting", t);
+
+                return true;
+            }
+
+            if (xmppConnection.isConnected())
+            {
+                xmppConnection.login();
```

Comment:
`login()` can potentially fail with an exception. Why not simplify to something like
```
try {
  if (!xmppConnection.isConnected())
     xmppConnnection.connect();
  xmppConnection.login();
}
```

Reply:
Cause we want to catch the connect error and it needs to be in try/catch

Reply:
aaa you mean if login fails to not try logging again ...

Reply:
Ok I will do that change

Reply:
I'm wondering why the `isConnected` check is  even there?